pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,274 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package offering
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+	"weak"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	awscache "github.com/aws/karpenter-provider-aws/pkg/cache"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/pricing"
+)
+
+type Provider interface {
+	InjectOfferings(context.Context, []*cloudprovider.InstanceType, *v1.EC2NodeClass, []string, *types.UID) []*cloudprovider.InstanceType
+}
+
+type DefaultProvider struct {
+	sync.RWMutex
+	snapshots map[types.UID]weak.Pointer[snapshot]
+
+	unavailableOfferings *awscache.UnavailableOfferings
+	pricingProvider      pricing.Provider
+}
+
+func NewDefaultProvider(unavailableOfferingsCache *awscache.UnavailableOfferings, pricingProvider pricing.Provider) *DefaultProvider {
+	return &DefaultProvider{
+		snapshots:            map[types.UID]weak.Pointer[snapshot]{},
+		unavailableOfferings: unavailableOfferingsCache,
+		pricingProvider:      pricingProvider,
+	}
+}
+
+func (p *DefaultProvider) InjectOfferings(
+	instanceTypes []*cloudprovider.InstanceType,
+	nodeClass *v1.EC2NodeClass,
+	allZones sets.Set[string],
+	snapshotUUID *types.UID,
+) []*cloudprovider.InstanceType {
+	subnetZones := lo.SliceToMap(nodeClass.Status.Subnets, func(s v1.Subnet) (string, string) {
+		return s.Zone, s.ZoneID
+	})
+	its := []*cloudprovider.InstanceType{}
+	for _, it := range instanceTypes {
```

Comment:
To confirm: We need to build things this way and we can't store the offerings with the instance types today because they might be out-of-date, so we effectively need to update the offerings each time we go through this process to make sure that we are getting fresh data every time -- do we have a feeling of the memory impact for this changes since the GetInstanceTypes() calls are effectively now more expensive to make

Reply:
It's more significant than I initially thought. We still need to live generate this for reserved offerings, but I'm thinking we can continue to cache all other offerings to reduce this as much as possible.

Reply:
We can -- although, I don't think you did in the latest change from what I saw